File: Editorial/0137_Single_Number_II/TypeScript/6_Boolean_Algebra_and_Karnaugh_Map.ts
function singleNumber(nums: number[]): number {
    // Count (modulo 3) bits
    let msb = 0,
        lsb = 0;
    // Process Every Num and update count bits
    for (let num of nums) {
        let new_lsb = (~msb & ~lsb & num) | (lsb & ~num);
        let new_msb = (lsb & num) | (msb & ~num);
        lsb = new_lsb;
        msb = new_msb;
    }
    // Return lsb as the answer
    return lsb;
}

File: Editorial/137_Single_Number_II/CPP/6_Boolean_Algebra_and_Karnaugh_Map.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int msb = 0, lsb = 0;
        for (int num : nums) {
            int new_lsb = (~msb & ~lsb & num) | (lsb & ~num);
            int new_msb = (lsb & num) | (msb & ~num);
            lsb = new_lsb;
            msb = new_msb;
        }
        return lsb;
    }
};
